src/platform_impl/linux/wayland/event_loop/mod.rs

```diff
@@ -62,12 +62,16 @@ impl<T: 'static> EventLoop<T> {
     pub fn new() -> Result<EventLoop<T>, Box<dyn Error>> {
         let connection = Connection::connect_to_env()?;
 
-        let (globals, event_queue) = globals::registry_queue_init(&connection)?;
+        let (globals, mut event_queue) = globals::registry_queue_init(&connection)?;
         let queue_handle = event_queue.handle();
 
         let event_loop = calloop::EventLoop::<WinitState>::try_new()?;
 
-        let winit_state = WinitState::new(&globals, &queue_handle, event_loop.handle())?;
+        let mut winit_state = WinitState::new(&globals, &queue_handle, event_loop.handle())?;
+
+        // NOTE: we must roundtrip with the server to ensure that the state is initialized
+        // correctly.
+        event_queue.roundtrip(&mut winit_state)?;
```

Comment:
Could you consider deferring the `NewEvents(Init)` and `Resumed` events until this round trip complete normally via the event loop?

This kind of synchronous round trip, similar to the round trip in `Window::new` seems like a kind of code smell that doesn't seem like it should be necessary ideally.

It could be good if the comment clarified, more specifically, what response is being waited for and why the event loop would be in an inconsistent state if it doesn't wait here.

Reply:
A roundtrip is basically a barrier for the server requests. It can't block indefinitely, because the `display.sync` is being sent under the hood.

Some compositors just have issues with "really async startup" here, so it's better to force a synchronization.

The one in window is a bit different though, it's to ensure that the `Window` was configured, given that you can't do anything with unconfigured one.

Reply:
That makes it sound like it's a workaround for fragile, maybe buggy, compositors.

That sounds potentially fair enough if that's the case but it still seems a shame that the comment doesn't point to a specific, verifiable, issue, like a workaround for a particular compositor, otherwise it sounds like a mysterious workaround that will end up being maintained indefinitely because noone can really verify what it fixes.

If it's more because the backend wants the response to fill in some `EventLoop` state (so it's a winit backend issue that means the event loop is configured asynchronously) (which the comment currently seems to imply) then I'd still wonder whether this wait could happen via the normal event loop, similar to how an `NSApp` gets 'launched' asynchronously, or an Android app resumes asynchronously and `NewEvents(Init)` will only be sent after the event loop has finished being set up asynchronously.

If possible it seems worthwhile to aim for only one standard place (the event loop proper) where the backend deals with I/O and the Wayland backend isn't the only backend that effectively wants to configure the `EventLoop` asynchronously.

Reply:
>That makes it sound like it's a workaround for fragile, maybe buggy, compositors.

Yes, And I also carry workarounds for GNOME for like 2 years, because it doesn't want fix spec violation. 

>That sounds potentially fair enough if that's the case but it still seems a shame that the comment doesn't point to a specific, verifiable, issue, like a workaround for a particular compositor, otherwise it sounds like a mysterious workaround that will end up being maintained indefinitely because noone can really verify what it fixes.

I'd add more clarifications for the places, where not really obvious, which were ported from the old code.

>If it's more because the backend wants the response to fill in some EventLoop state (so it's a winit backend issue that means the event loop is configured asynchronously) (which the comment currently seems to imply) then I'd still wonder whether this wait could happen via the normal event loop, similar to how an NSApp gets 'launched' asynchronously, or an Android app resumes asynchronously and NewEvents(Init) will only be sent after the event loop has finished being set up asynchronously.

This roundtrip is not for winit, but for compositor state, nothing winit should do here, it's required only on initial loop creation, because you bind globals only once.

>) then I'd still wonder whether this wait could happen via the normal event loop, similar to how an NSApp gets 'launched' asynchronously, or an Android app resumes asynchronously and NewEvents(Init) will only be sent after the event loop has finished being set up asynchronously.

Be aware that it's a short wait (unless your compositor is dead broken) and it's required once per event loop creation, so the `::new` is the only place for it. Nothing else is required here, it can't even send events for the user.

>If possible it seems worthwhile to aim for only one standard place (the event loop proper) where the backend deals with I/O and the Wayland backend isn't the only backend that effectively wants to configure the EventLoop asynchronously.

We don't configure anything in the winit's event loop it's already done. The roundtrip is to syncchorinize with the compositor, so it won't get confused for whatever reason it could get confused, there's no state broken state in winit before/after roundtrip.

The bug it was solving was purely to "compositor for whatever reason still things that the object is not alive", because it's likely a race in it, so I've decided for a more conservative startup, which was even before(but was hidden inside the macros).